db/hints/sync_point.cc

```diff
@@ -67,16 +69,30 @@ static std::vector<sync_point::shard_rps> decode_one_type_v1(uint16_t shard_coun
     return ret;
 }
 
+static uint64_t get_checksum(signed char* data, size_t size) {
+    xx_hasher h;
+    h.update(reinterpret_cast<const char*>(data), size);
+    return h.finalize_uint64();
+}
+
 sync_point sync_point::decode(sstring_view s) {
     bytes raw = base64_decode(s);
     if (raw.empty()) {
         throw std::runtime_error("Could not decode the sync point - not a valid hex string");
     }
-    if (raw[0] != 1) {
+
+    if (raw[0] == 2) {
+        seastar::simple_memory_input_stream in_checksum{reinterpret_cast<const char*>(raw.end()) - sizeof(uint64_t), sizeof(uint64_t)};
+        uint64_t checksum = ser::serializer<uint64_t>::read(in_checksum);
+        if (checksum != get_checksum(raw.data(), raw.size() - sizeof(uint64_t))) {
```

Comment:
If you apply my previous suggestion about changing the interface of `get_checksum`, you can use the [`.substr()`](https://en.cppreference.com/w/cpp/string/basic_string_view/substr) method (`sstring_view` an alias over `std::string_view`) which you can use to create a subslice of the original `string_view`:

```c++
if (checksum != calculate_checksum(raw.substr(0, raw.size() - checksum_size)) {
    // ...
}
```

Reply:
Unfortunately, it does not work, because `raw` has a type `bytes`.  Some casting is necessary here. I've made it work, but I don't think it's very clear.